### PR TITLE
Rewrite the Kademlia k-buckets to be more explicit

### PR DESCRIPTION
**Pull request by tomaka**
Right now the API of k-buckets is pretty much magic. There are tons of corner cases that aren't covered.
This PR makes the API clearer so that the behaviour behaves properly.

Work-in-progress because this needs a review pass from me, and because more tests are needed.

Also we probably want to make time management from the k-buckets part of the API, for better testability and for WASM.

### CODE REVIEW

**Comment by tomaka** [comment]
Forgot to mention that this was based on top of #994. Only the last commit is relevant to this PR.

**Comment by tomaka** [comment]
Ready for review!

I also added a `KBucketAdded` event that notifies when we insert a new value in a bucket.

File: protocols/kad/src/kbucket.rs [hunk 1]
```diff
@@ -161,254 +176,502 @@ where
...
+        // Update the pending node state.
+        // TODO: must be reported to the user somehow, in a non-annoying API
+        if let Some(pending) = self.tables[bucket_num].pending_node.take() {
+            if pending.replace < Instant::now() {
+                let table = &mut self.tables[bucket_num];
+                let first_connected_pos = &mut table.first_connected_pos;
+                // If all the nodes in the bucket are connected, then there shouldn't be any
+                // pending node.
+                debug_assert!(*first_connected_pos >= 1);
+                let _ = table.nodes.remove(0);
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
                table.nodes.remove(0);
```

File: protocols/kad/src/kbucket.rs [hunk 2]
```diff
@@ -161,254 +176,502 @@ where
...
+
+impl<'a, TPeerId, TVal> EntryNotInKbucket<'a, TPeerId, TVal>
+where
+    TPeerId: KBucketsPeerId + Clone,
+{
+    /// Inserts the node as connected, if possible.
+    pub fn insert_connected(self, value: TVal) -> InsertOutcome<TPeerId> {
+        let table = {
+            let num = self.parent.bucket_num(&self.peer_id)
+                .expect("we can only build a EntryInKbucketDisc if we know of a bucket; QED");
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
                .expect("we can only build a EntryNotInKbucket if we know of a bucket; QED");
```

File: protocols/kad/src/kbucket.rs [hunk 3]
```diff
@@ -161,254 +176,502 @@ where
...
-                intermediate.insert(pos, self.my_id.clone());
-            }
+    /// Inserts the node as disconnected, if possible.
+    ///
+    /// > **Note**: This function will never return `Pending`. If the bucket is full, we simply
+    /// >           do nothing.
+    pub fn insert_disconnected(self, value: TVal) -> InsertOutcome<TPeerId> {
+        let table = {
+            let num = self.parent.bucket_num(&self.peer_id)
+                .expect("we can only build a EntryInKbucketDisc if we know of a bucket; QED");
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
                .expect("we can only build a EntryNotInKbucket if we know of a bucket; QED");
```

File: protocols/kad/src/kbucket.rs [hunk 4]
```diff
@@ -161,254 +176,502 @@ where
...
-        table.flush(self.unresponsive_timeout);
+impl<'a, TPeerId, TVal> EntryInKbucketDiscPending<'a, TPeerId, TVal>
+where
+    TPeerId: KBucketsPeerId + Clone,
+{
+    /// Returns the value associated to the entry in the bucket.
+    pub fn value(&mut self) -> &mut TVal {
+        let table = {
+            let num = self.parent.bucket_num(&self.peer_id)
+                .expect("we can only build a EntryInKbucketPending if we know of a bucket; QED");
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
                .expect("we can only build a EntryInKbucketDiscPending if we know of a bucket; QED");
```

File: protocols/kad/src/kbucket.rs [hunk 5]
```diff
@@ -161,254 +176,502 @@ where
...
-                let elem = table.nodes.remove(pos);
-                table.first_connected_pos -= 1;
-                table.nodes.insert(table.first_connected_pos, elem);
-            }
-            table.latest_update = Instant::now();
-            Update::Updated
+        assert!(table.pending_node.as_ref().map(|n| &n.node.id) == Some(self.peer_id));
+        &mut table.pending_node
+            .as_mut()
+            .expect("we can only build a EntryInKbucketPending if the node is pending; QED")
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
            .expect("we can only build a EntryInKbucketDiscPending if the node is pending; QED")
```

File: protocols/kad/src/kbucket.rs [hunk 6]
```diff
@@ -161,254 +176,502 @@ where
...
-            // to the oldest node.
-            let pending_node = Node {
-                id: id.clone(),
-                value: Default::default(),
+    /// Reports that we are now connected to the given node.
+    pub fn set_connected(self) -> EntryInKbucketConnPending<'a, TPeerId, TVal> {
+        {
+            let table = {
+                let num = self.parent.bucket_num(&self.peer_id)
+                    .expect("we can only build a EntryInKbucketConnPending if we know of a bucket; QED");
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
                    .expect("we can only build a EntryInKbucketDiscPending if we know of a bucket; QED");
```

File: protocols/kad/src/kbucket.rs [hunk 7]
```diff
@@ -161,254 +176,502 @@ where
...
+                &mut self.parent.tables[num]
             };
-            table.pending_node = Some((pending_node, Instant::now()));
-            Update::Pending(&table.nodes[0].id)
 
-        } else {
-            debug_assert!(table.first_connected_pos == 0 || table.pending_node.is_some());
-            Update::Discarded
+            let mut pending = table.pending_node.as_mut()
+                .expect("we can only build a EntryInKbucketConnPending if there's a pending node; QED");
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]:
```suggestion
                .expect("we can only build a EntryInKbucketDiscPending if there's a pending node; QED");
```

File: protocols/kad/src/kbucket.rs [hunk 8]
```diff
@@ -161,254 +176,502 @@ where
...
+    ///
+    /// Pending nodes are ignored.
+    pub fn find_closest(&mut self, id: &impl KBucketsPeerId<TPeerId>) -> VecIntoIter<TPeerId> {
+        // TODO: optimize
+        let mut out = Vec::new();
+        for table in self.tables.iter_mut() {
+            for node in table.nodes.iter() {
+                out.push(node.id.clone());
+            }
+            // TODO: that's kind of wrong
```
Comment (twittner) [inline protocols/kad/src/kbucket.rs]: Could you be more specific?

Comment (tomaka) [reply protocols/kad/src/kbucket.rs]: We need the same logic here as in the `entry` method. The code here works but doesn't have the same behaviour, which is a bit stupid.

In practice though, we need a way to report when a pending node replaces a valid node (there's another TODO for that), meaning that we need to rework this anyway.

Comment (twittner) [reply protocols/kad/src/kbucket.rs]: Maybe this should be mentioned in the comment?

Comment (tomaka) [reply protocols/kad/src/kbucket.rs]: Done

**Comment by tomaka** [comment]
We discussed this a bit in private.
Whereas I don't trust the code that it's bug free, I have high confidence that this PR is a huge improvement over the current situation.
There's an `unimplemented!` in `behaviour.rs` which I intend to fix before we publish 0.6.
